docs/rfcs/020-pageserver-s3-coordination.md

```diff
@@ -0,0 +1,75 @@
+# Coordinating access of multiple pageservers to the same s3 data
+
+## Motivation
+
+There are some blind spots around coordinating access of multiple pageservers to the same s3 data.
+Currently this is applicable only to tenant relocation case, but in the future we'll need to solve similar problems for replica/standby pageservers.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Pageserver
+
+## The problem:
+
+### Relocation
+
+During relocation both pageservers can write to s3. This should be ok for all data except the `index_part.json`. For index part it causes problems during compaction/gc because they remove files from index/s3.
+
+Imagine this case:
+
+```mermaid
+sequenceDiagram
+    autonumber
+    participant PS1
+    participant S3
+    participant PS2
+
+    PS1->>S3: Uploads L1, L2 <br/> Index contains L1 L2
+    PS2->>S3: Attach called, sees L1, L2
+    PS1->>S3: Compaction comes <br/> Removes L1, adds L3
+    note over S3: Index now L2, L3
+    PS2->>S3: Uploads new layer L4 <br/> (added to previous view of the index)
+    note over S3: Index now L1, L2, L4
+```
+
+At this point it is not possible to restore from index, it contains L2 which is no longer available in s3 and doesnt contain L3 added by compaction by the first pageserver. So if any of the pageservers restart initial sync will fail (or in on-demand world it will fail a bit later during page request from missing layer)
+
+### Standby pageserver
+
+Another related case is standby pageserver. In this case second pageserver can be used as a replica to scale reads and serve as a failover target in case first one fails.
```

Comment:
This seems like a cool feature, however, my question is generally:
- are pageservers read-bound, and if so what amount of read scaling are we expecting to get out of this. I'd be interested in weighing the cost of multiplying the cost of pageservers against the relative benefit.
- do the read and write workloads interact or impact each other?

You could imagine alternate solutions whereby you have more smaller pageservers with fewer numbers of tenants that you then scale up and down (vertically) as needed to accommodate load.

Reply:
Yeah, that needs a lot of experiments. I think this is rather out of scope for the RFC. It is more about how to allow multiple pageservers look into the same s3 prefix without corrupting things underneath.

> are pageservers read-bound

In read heavy workloads I think they are. This way reads can scale horizontally, so theoretically two pageservers can give twice the throughput. But then comes work from the compute side, how to properly schedule requests, ongoing pre-fetching work, etc etc. So thats hard to guess what will be the real scaling factor

> do the read and write workloads interact or impact each other

Since pageserver is multitenant there can be some skewed picture overall if one tenant's workload is read heavy, and for the other one it is write heavy

> smaller pageservers with fewer numbers of tenants

Especially interesting to consider edge deployments, where you have one writer and many readers across the globe. For this case having even small pageservers but close to the client looks like a good way to cut latency